### DISCUSSION TITLE
How about adding `IBytes` and `IBencodex`?

### DISCUSSION THREAD

**Discussion by greymistcube** [planetarium/libplanet#2560] [Ideas]
There are numerous entities in [libplanet] using `ImmutableArray<byte>`, `byte[]`, and/or `IValue`, most, if not all, of which have a canonical representation where conversion between the three are done quite frequently. However, without an overarching `interface`, there are several problems:
- Lack of standard implementation. Each `class` or `struct` wanting to have `ImmutableArray<byte>`, `byte[]`, and/or `IValue` representation can have wildly different property/method names.
- As the user is unsure whether there is canonical representation already associated with a data type, it is more prone to mistakes.
- Frequent use of local `Codec` instantiation.
- Double+ conversion between `ImmutableArray<byte>` and `byte[]`.

I'm thinking something along the lines of the following:
```csharp
public interface IBytes
{
    public ImmutableArray<byte> ImmutableBytes { get; } // ImBytes? ByteArray?
    public byte[] Bytes { get; }
}

public interface IBencodex : IBytes
{
    public Bencodex.Types.IValue Bencodexed { get; }    // BxEncoded? Encoded? Serialized?
}
```

I'm not satisfied with the names tho. 🙄
I forgot where `ToByteArray()` and `ByteArray` were originally derived from, even though I am already accustomed to it, but I personally think it is really confusing to have `ImmutableArray<byte> ByteArray` and `byte[] ToByteArray()` as they are currently being used.

[libplanet]: https://github.com/planetarium/libplanet

**Comment by ilgyu-yi** [comment]
Closing, since now we have `IBencodable`.
(Though we still does not have interface like `IBytes`, we can generate `byte[]` with `Bencodex.Codec.Encode(IBencodable.Bencoded)`).
- https://github.com/planetarium/libplanet/pull/2778
